---
params:
  projectdir: null # Use when loading. If null, here::here() will be used. Can also hard code if for some reason you need that
  project_title: "GeneTox21"    # Change depending on your project name
  project_name: "Your name here" 
  project_description: null # description of project for report, optional
  analysis_name: "default" # short string identifying analysis settings. This will appear in the analysis directory and file names
  batch_var: "batch"          # For example, multiple plates
  dose: "dose"               # If there is a dose in the experiment; otherwise use NULL
  platform: "TempO-Seq"      # TempO-Seq Or RNA-Seq
  nmr_threshold: 100000    # 10% of 1M reads for TempOSeq = 100,000; 10% of 10M reads for RNA-Seq = 1,000,000.
  write_additional_output: FALSE # export BMD, biomarker, and biosets output? facet variable should contain chemical and timepoint information
  celltype: "MCF7 cells" # only required for biosets output. Cell type or species used
  units: "uM" # only required for biosets output. Units of dose
  species: "human"                 # one of human, mouse, rat, hamster
  design: "group"                # single experimental group of interest; entries in this column must match the contrast names.
  intgroup: ["group"]            # experimental group of interest plus covariates; can be more than on
  intgroup_to_plot: ["group"]   # for PCA plots, add tabs for one or more groups to color
  formula_override: null # e.g., "~batch + condition", to enable users to specify the DESeq2 formula to use
  deseq_facet: "chemical"       # If you have many different experimental groups, you may subset the report by specifying a column in the metadata to filter groups, and then setting the group of interest in deseq_filter
  deseq_filter:  null        # Which group will this report be done on?
  reports_facet: null #  
  reports_filter:  null # 
  sortcol: "dose"                # Optionally, a column by which to sort the contrasts
  solvent_control: "solvent_control"
  lenient_contrasts: FALSE         # Use either column (exp, cont) of contrasts file to limit what is included in the report (instead of just exp column)
  strict_contrasts: FALSE         # Use BOTH columns (exp, cont) of contrasts file to limit what is included in the report
  exclude_samples: null      # Optionally, a vector of sample names to exclude from the analysis
  exclude_groups: null       # Optionally, a vector of groups to exclude from the analysis. By default this is assumed to be in the column specified by params$design.
  include_only_column:  null # Restrict analysis to group(s) in the column listed here based on params$include_only_group.
  include_only_group:  null  # Restrict analysis to this/these group(s) within the column listed in params$include_only_column
  cpus: 41                       # Set to a lower number (e.g., 2 to 4) if you aren't working in a server environment
  run_pathway_analysis: TRUE     # Optionally disable pathway analysis if not available for your organism
  wikipathways_directory: "~/shared/dbs/wikipathways"
  linear_fc_filter_DEGs: 1.5 # Default 1.5
  linear_fc_filter_biosets: 1.2 # Default 1.2. Used for biosets output when write_additional_output=True
  biospyder_dbs: "~/shared/dbs/biospyder/"
  biospyder_manifest_file:  "181019_Human_S1500_Surrogate_1.2_Manifest.txt"
  # "191113_Human_S1500_Surrogate_2.0_Manifest.csv"
  # "181019_Human_S1500_Surrogate_1.2_Manifest.txt"
  # "191004_Human_Whole_Transcriptome_2.0_Manifest.txt"
  wikipathways_filename: "wikipathways-20210810-gmt-Homo_sapiens.gmt"
  KEGGpathways_filename: "c2.cp.kegg_medicus.v2023.2.Hs.symbols.gmt"
  # "wikipathways-20210810-gmt-Mus_musculus.gmt"
  # "wikipathways-20210810-gmt-Rattus_norvegicus.gmt"
  nBestFeatures: 20              # The number of best features to make plots of their counts
  nBest: 100                     # Number of features to include in table and limiting PCA/clustering analysis
  nHeatmap: 50                   # Number of most variable genes for heatmap
  nHeatmapDEGs: 50               # Number of DEGs for heatmap
  cooks: FALSE                   # the DESeq Cook's distance cutoff, or FALSE to disable it
  filter_gene_counts: FALSE      # Filter genes to those with at least 1 count in 1 sample. Disabled by default for biomarker file output. Can enable to increase speed of analysis
  generate_main_report: TRUE
  generate_stats_report: TRUE
  generate_data_explorer_report: TRUE
  generate_go_pathway_report: TRUE
  output_digits: 5
  parallel: FALSE
  species_data: null
  count_data_file: null
  sampledata_sep: null
  MinCount: null
  alpha: 0.05
  feature_id: null
  biospyder: null
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    code_folding: hide
    theme: spacelab           # flatly spacelab sandstone cerulean
    code_download: true
---

```{r docSetup, include = FALSE, warning = FALSE, message = FALSE}
#### Record start time
startTime <- Sys.time()
library('DT')
library('plotly')
library('crosstool')
library('crosstalk')
library('R.ODAF.utils')

db <- AnnotationDbi::loadDb(params$species_data$orgdb)

# now that we have all our params etc loaded, we can set the HTML title of our document 
```
---
title: "`r params$platform` DEG analysis - data explorer"
---


# `r params$project_title` - Data explorer {-}

Date report generated: `r format(Sys.time(), '%d %B, %Y')`  

Report prepared by: `r params$project_bioinformatics_name`

Report prepared for: `r params$project_name`


***

This report contains detailed tables of DEG data.

```{r load_facet_data, include=FALSE}
load_facet_data(paths, params)
```


This table shows the significant DEGs (passing all filtering criteria) ordered by their absolute fold change. Use the search function to find your feature of interest or sort by one of the columns. You can limit to a single contrast if desired.

The links will search the NCBI and Ensembl sites for the gene of interest. Note that for genes without designated symbols, Ensembl usually has a result while NCBI does not.

```{r 'topFeatures', results = 'asis', warning = FALSE, message = FALSE}
human_rounding <- 2
searchURL_NCBI <- "http://www.ncbi.nlm.nih.gov/gene/?term="
searchURL_Ensembl <- "https://ensembl.org/Multi/Search/Results?q="
## Add search url if appropriate
res.df.dt <- allResults %>%
  dplyr::filter(abs(linearFoldChange) > params$linear_fc_filter_DEGs) %>%
  dplyr::filter(padj < params$alpha) %>%
  arrange(-abs(linearFoldChange))

descriptions <- AnnotationDbi::select(db,
                                      columns = c("ENSEMBL", "GENENAME"),
                                      keys = res.df.dt$Ensembl_Gene_ID,
                                      keytype="ENSEMBL") %>%
  distinct()
colnames(descriptions) <- c("Ensembl_Gene_ID","description")
res.df.dt <- res.df.dt %>% left_join(descriptions)

res.df.dt$URLs <- paste0('<a href="',
                         searchURL_NCBI,
                         res.df.dt$Ensembl_Gene_ID,
                         '" rel="noopener noreferrer" target="_blank">',
                         'NCBI</a><br/>',
                         '<a href="',
                         searchURL_Ensembl,
                         res.df.dt$Ensembl_Gene_ID,
                         '" rel="noopener noreferrer" target="_blank">',
                         'Ensembl</a>')

res.df.dt <- res.df.dt %>% dplyr::relocate(URLs)

res.df.dt$Gene <- paste0(res.df.dt$Ensembl_Gene_ID,
                         '<br/>',
                         res.df.dt$Gene_Symbol)

res.df.dt <- res.df.dt %>% dplyr::relocate(Gene)


res.df.dt[, 'padj'] <- format(res.df.dt[, 'padj'],
                              scientific = TRUE,
                              digits = human_rounding)
res.df.dt[, 'pvalue'] <- format(res.df.dt[, 'pvalue'],
                                scientific = TRUE,
                                digits = human_rounding)
DT::datatable(res.df.dt,
          options = list(pagingType = 'full_numbers',
                         pageLength = 20,
                         scrollX = '100%',
                         dom = 'Bfrtip',
                         buttons = c('copy',
                                     'csv',
                                     'excel',
                                     'pdf',
                                     'print',
                                     'colvis'),
                         columnDefs = list(list(visible = FALSE, targets = c(2,3, 4, 5, 7, 8, 9)))),
          escape = FALSE,
          extensions = 'Buttons',
          rownames = FALSE,
          filter = "top",
          ) %>% 
  DT::formatRound(which(!colnames(res.df.dt) %in% c('pvalue',
                                                    'padj',
                                                    'Feature',
                                                    'contrast',
                                                    'description',
                                                    'Gene',
                                                    'Gene_Symbol',
                                                    'URLs',
                                                    'Ensembl_Gene_ID',
                                                    'Feature_ID')),
                  human_rounding)


```

# Interactive Gene-level plot

This section allows the normalized counts of specific genes (or probes, in the case of TempO-Seq) to be viewed.

Note that for ease of use, only the top 1000 genes (by absolute fold change) are included.

```{r geneplots, warning = FALSE, message = FALSE}

# Make a table of counts
counts_by_group <- as.data.frame(counts(dds, normalized = TRUE))
counts_by_group <- as.data.frame(counts_by_group)

# counts_by_group <- log2(counts_by_group + 1) # Optional
# Add gene name
counts_by_group$gene <- row.names(counts_by_group)
# Clean up table, make into long format for plotting
counts_by_group <- counts_by_group %>%
  tidyr::pivot_longer(!gene,
               names_to = "Sample",
               values_to = "Normalized Counts")

# Annotate table with experimental group of interest (params$design)
counts_by_group <- counts_by_group %>%
  left_join(exp_metadata %>% dplyr::select(c(!!sym(params$design),original_names)),
            by = c("Sample" = "original_names"))
names(counts_by_group)[4] <- "group"

# Filter by absolute fold change to include a reasonable number of genes
genesToSelect <- allResults %>%
  dplyr::arrange(-abs(log2FoldChange)) %>%
  head(1000) # Seems to be a reasonable maximum...

# Order by alphabet to help readability in dropdown menu
genesAlphabetized <- genesToSelect %>%
  dplyr::arrange(-abs(log2FoldChange)) %>%
  dplyr::arrange(Gene_Symbol)

# Make sure genes are unique to avoid replicated rows and points in plot
genesAlphabetized <- genesAlphabetized %>%
  dplyr::distinct(Feature_ID, .keep_all = TRUE)

# Get the data 
counts_by_group <- counts_by_group %>%
  dplyr::filter(gene %in% genesAlphabetized[["Feature_ID"]]) %>%
  dplyr::left_join(genesAlphabetized, by=c(gene="Feature_ID")) %>%
  dplyr::select(gene,Sample,'Normalized Counts',group,Gene_Symbol)

# Create "shared" data object - this is for enabling live filtering
counts_by_group_shared <- SharedData$new(counts_by_group %>%
  mutate(group = factor(group, levels = mixedsort(unique(group)))))

# Make an initial filter using the top gene
top_gene = genesToSelect[["Feature_ID"]][1]
i = as.character(which(counts_by_group$Gene_Symbol == top_gene))

# Make a "transceiver" to set initial state of filter
tx = crosstool(counts_by_group_shared,
               "transceiver",
               init = i,
               channel = "filter",
               reset = rownames(counts_by_group))
# Plot interactive plotly with filter on gene name
bscols(widths = c(2,NA),
  list(
    filter_select("gene_to_filter",
                  "Gene",
                  counts_by_group_shared,
                  ~Gene_Symbol,
                  unique(genesAlphabetized[["Feature_ID"]]),
                  multiple = F)
  ),
  list(
    plot_ly(data = counts_by_group_shared,
            x = ~group,
            y = ~`Normalized Counts`,
            color = ~group,
            type = "box",
            boxpoints = "all",
            jitter = 0.6,
            pointpos = 0,
            text = ~Sample,
            boxmean = T) %>%
      layout(xaxis = list(title = ''), 
             yaxis = list(title = 'Normalized Counts'),
             legend = list(title = list(text = '<b> Experimental Group </b>'))) %>%
      add_annotations(
        text = "Gene",
        x = 0.5, y = 1,
        xref = "paper", yref = "paper",
        xanchor = "center", yanchor = "bottom",
        showarrow = FALSE
      ),
    tx)
)


```

```{r disconnect, include = F}
DBI::dbDisconnect(dbconn(db))
```

```{r, child=file.path(paths$root,'Rmd','session_info.Rmd')}
```