---
params:
  projectdir: null # Use when loading. If null, here::here() will be used. Can also hard code if for some reason you need that
  project_title: "Default"    # Change depending on your project name
  project_name: "Your name here" 
  project_description: null # description of project for report, optional
  analysis_name: "default" # short string identifying analysis settings. This will appear in the analysis directory and file names
  batch_var: "batch"          # For example, multiple plates
  dose: "dose"               # If there is a dose in the experiment; otherwise use NULL
  platform: "TempO-Seq"      # TempO-Seq Or RNA-Seq
  nmr_threshold: 100000    # 10% of 1M reads for TempOSeq = 100,000; 10% of 10M reads for RNA-Seq = 1,000,000.
  write_additional_output: FALSE # export BMD, biomarker, and biosets output? facet variable should contain chemical and timepoint information
  celltype: "MCF7 cells" # only required for biosets output. Cell type or species used
  units: "uM" # only required for biosets output. Units of dose
  species: "human"                 # one of human, mouse, rat, hamster
  design: "group"                # single experimental group of interest; entries in this column must match the contrast names.
  intgroup: ["group"]            # experimental group of interest plus covariates; can be more than one
  intgroup_to_plot: ["group"]   # for PCA plots, add tabs for one or more groups to color
  formula_override: null # e.g., "~batch + condition", to enable users to specify the DESeq2 formula to use
  deseq_facet: "chemical"       # If you have many different experimental groups, you may subset the report by specifying a column in the metadata to filter groups, and then setting the group of interest in deseq_filter
  deseq_filter:  null        # Which group will this report be done on?
  reports_facet: null #  
  reports_filter:  null # 
  sortcol: "dose"                # Optionally, a column by which to sort the contrasts
  solvent_control: "solvent_control"
  lenient_contrasts: FALSE         # Use either column (exp, cont) of contrasts file to limit what is included in the report (instead of just exp column)
  strict_contrasts: FALSE         # Use BOTH columns (exp, cont) of contrasts file to limit what is included in the report
  exclude_samples: null      # Optionally, a vector of sample names to exclude from the analysis
  exclude_groups: null       # Optionally, a vector of groups to exclude from the analysis. By default this is assumed to be in the column specified by params$design.
  include_only_column:  null # Restrict analysis to group(s) in the column listed here based on params$include_only_group.
  include_only_group:  null  # Restrict analysis to this/these group(s) within the column listed in params$include_only_column
  cpus: 41                       # Set to a lower number (e.g., 2 to 4) if you aren't working in a server environment
  run_pathway_analysis: TRUE     # Optionally disable pathway analysis if not available for your organism
  wikipathways_directory: "~/shared/dbs/wikipathways"
  linear_fc_filter_DEGs: 1.5 # Default 1.5
  linear_fc_filter_biosets: 1.2 # Default 1.2. Used for biosets output when write_additional_output=True
  biospyder_dbs: "~/shared/dbs/biospyder/"
  biospyder_manifest_file:  "181019_Human_S1500_Surrogate_1.2_Manifest.txt"
  # "191113_Human_S1500_Surrogate_2.0_Manifest.csv"
  # "181019_Human_S1500_Surrogate_1.2_Manifest.txt"
  # "191004_Human_Whole_Transcriptome_2.0_Manifest.txt"
  wikipathways_filename: "wikipathways-20210810-gmt-Homo_sapiens.gmt"
  KEGGpathways_filename: "c2.cp.kegg_medicus.v2023.2.Hs.symbols.gmt"
  # "wikipathways-20210810-gmt-Mus_musculus.gmt"
  # "wikipathways-20210810-gmt-Rattus_norvegicus.gmt"
  nBestFeatures: 20              # The number of best features to make plots of their counts
  nBest: 100                     # Number of features to include in table and limiting PCA/clustering analysis
  nHeatmap: 50                   # Number of most variable genes for heatmap
  nHeatmapDEGs: 50               # Number of DEGs for heatmap
  cooks: FALSE                   # the DESeq Cook's distance cutoff, or FALSE to disable it
  filter_gene_counts: FALSE      # Filter genes to those with at least 1 count in 1 sample. Disabled by default for biomarker file output. Can enable to increase speed of analysis
  generate_main_report: TRUE
  generate_stats_report: TRUE
  generate_data_explorer_report: TRUE
  generate_go_pathway_report: TRUE
  generate_tgxddi_report: TRUE
  output_digits: 5
  parallel: FALSE
  species_data: null
  count_data_file: null
  sampledata_sep: null
  MinCount: null
  alpha: 0.05
  feature_id: null
  biospyder: null
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    code_folding: hide
    theme: spacelab           # flatly spacelab sandstone cerulean
    code_download: true
bibliography: "`r file.path(params$projectdir, 'Rmd/references.bib')`"

---


```{r docSetup, warning = FALSE, message = FALSE, include = F}
#### Record start time
startTime <- Sys.time()

library(kableExtra)
library(pheatmap)
library(here)
library(crosstalk)
library(crosstool)
library(plotly)
library(tidytext)
library(DESeq2)
library(dplyr)
library(tibble)

db <- AnnotationDbi::loadDb(params$species_data$orgdb)
```


```{r load_functions, include=F}

has_description <- !is.na(params$project_description)
description_text <- paste("Purpose of report:", params$project_description)

# now that we have all our params etc loaded, we can set the HTML title of our document 
```
---
title: "`r params$platform` DEG analysis - main report"
---


# `r params$project_title` - RNA-seq analysis {-}

Date report generated: `r format(Sys.time(), '%d %B, %Y')`  

Report prepared by: `r params$project_bioinformatics_name`

Report prepared for: `r params$project_name`

`r if(has_description){description_text}`

***

This report is meant to help explore DESeq2 results and was generated using RMarkdown.

This main report displays the overall results of the experiment and contains the most commonly used visualizations. Additional statistics and graphs are available in the included "Extra stats report", and a "Data explorer report" with interactive tables and graphs to investigate specific genes is also available. Finally, the GO and KEGG pathway analysis results are contained in the "GO and pathway analysis report"

The structure of this document originally extended the DESeq2 regionReport RMarkdown (Collado-Torres, Jaffe, and Leek 2017) but has been substantially reworked since then.

Code sections are hidden by default but can be viewed by clicking the 'code' button.


```{r load_facet_data, include=FALSE}
load_facet_data(paths, params)

if (params$platform == "TempO-Seq") {
  id_table <- params$biospyder %>% dplyr::select(Feature_ID = Probe_Name, Gene_Symbol, Ensembl_Gene_ID)
} else {
  id_table <- AnnotationDbi::select(db,
                                    columns = c("ENSEMBL", "SYMBOL"),
                                    keys = allResults$Feature_ID, keytype="ENSEMBL") %>%
    distinct() %>%
    mutate(Gene_Symbol = coalesce(SYMBOL, ENSEMBL)) %>%
    dplyr::select(Feature_ID = ENSEMBL, Ensembl_Gene_ID = ENSEMBL, Gene_Symbol)
}
```

# Volcano plots

## All data

```{r volcano_plot_all, fig.width = 8, fig.height = 6, warning = TRUE, collapse = TRUE}

ggplot(allResults, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(size = 0.5, alpha = 0.3) +
  geom_point(data = significantResults, size = 0.5, alpha = 0.9, color = "red") +
  geom_vline(xintercept = c(-log2(1.5), log2(1.5)), color = "red", alpha = 1.0) + 
  geom_hline(yintercept = -log10(0.05), color = "blue", alpha = 1.0) +
  scale_x_continuous(name = "log2 Fold Change", limits = c(-5, 5)) +
  scale_y_continuous(name = "-log10 adjusted p-value", limits = c(0, 6)) + 
  facet_wrap(~contrast, ncol = 4, labeller = labeller(
    .default = label_wrap_gen(15))) +
  theme_bw()
```

## By contrast

Select a contrast from the dropdown list to filter by contrast. Note that this plot is turned off by default for RNA-seq analysis.
```{r volcano_by_contrast, fig.width = 10, fig.height = 6, warning = TRUE, collapse = TRUE, eval = !(params$platform == "RNA-Seq")}
volcano_df <-  as.data.frame(allResults) %>%
    dplyr::select(c("Feature_ID",
                    "Gene_Symbol",
                    "log2FoldChange",
                    "linearFoldChange",
                    "padj",
                    "contrast")) %>%
    distinct()
volcano_df_sig <- as.data.frame(significantResults)

volcano_status_FC <- paste0("Linear fold change < ",params$linear_fc_filter_DEGs)
volcano_status_passed <- "Passed all filters"
volcano_status_RODAF <- "Filtered by R-ODAF criteria"
volcano_status_FDR <- paste0("FDR >= ", params$alpha)

volcano_df <- volcano_df %>%
  mutate(
    sig_group = case_when(
      padj < params$alpha & abs(linearFoldChange) <= params$linear_fc_filter_DEGs                                             ~ volcano_status_FC,
      padj < params$alpha & abs(linearFoldChange) > params$linear_fc_filter_DEGs & Feature_ID %in% volcano_df_sig$Feature_ID  ~ volcano_status_passed,
      padj < params$alpha & abs(linearFoldChange) > params$linear_fc_filter_DEGs & !(Feature_ID %in% volcano_df_sig$Feature_ID)  ~ volcano_status_RODAF,
      TRUE                                                                                                               ~ volcano_status_FDR
    )
  )

status_colors <- list(FDR = "#66c2a5", FC = "#8da0cb", passed = "#e78ac3", RODAF = "#fc8d62")
all_statuses <- list(FDR = volcano_status_FDR, FC = volcano_status_FC, passed = volcano_status_passed, RODAF = volcano_status_RODAF)
all_status_keys <- c("FDR", "FC", "passed", "RODAF")
volcano_contrasts <- unique(allResults$contrast)

x <- list()
y <- list()
volcano_labs <- list()
for (cont in volcano_contrasts) {
  x[[cont]] <- list(
    FDR = subset(volcano_df, contrast == cont & sig_group == volcano_status_FDR) %>% pull(log2FoldChange),
    FC = subset(volcano_df, contrast == cont & sig_group == volcano_status_FC) %>% pull(log2FoldChange),
    passed = subset(volcano_df, contrast == cont & sig_group == volcano_status_passed) %>% pull(log2FoldChange),
    RODAF = subset(volcano_df, contrast == cont & sig_group == volcano_status_RODAF) %>% pull(log2FoldChange)
  )
  y[[cont]] <- list(
    FDR = subset(volcano_df, contrast == cont & sig_group == volcano_status_FDR) %>% mutate(log10p = -log10(padj)) %>% pull(log10p),
    FC = subset(volcano_df, contrast == cont & sig_group == volcano_status_FC) %>% mutate(log10p = -log10(padj)) %>% pull(log10p),
    passed = subset(volcano_df, contrast == cont & sig_group == volcano_status_passed) %>% mutate(log10p = -log10(padj)) %>% pull(log10p),
    RODAF = subset(volcano_df, contrast == cont & sig_group == volcano_status_RODAF) %>% mutate(log10p = -log10(padj)) %>% pull(log10p)
  )
  volcano_labs[[cont]] <- list(
    FDR = subset(volcano_df, contrast == cont & sig_group == volcano_status_FDR) %>% pull(Gene_Symbol),
    FC = subset(volcano_df, contrast == cont & sig_group == volcano_status_FC) %>% pull(Gene_Symbol),
    passed = subset(volcano_df, contrast == cont & sig_group == volcano_status_passed) %>% pull(Gene_Symbol),
    RODAF = subset(volcano_df, contrast == cont & sig_group == volcano_status_RODAF) %>% pull(Gene_Symbol)
  )
}


fig <- plot_ly(type = 'scatter', mode='markers')
for(cont in volcano_contrasts){
  for (status in all_status_keys){
    fig <- fig %>% add_trace(x=x[[cont]][[status]],
                             y=y[[cont]][[status]],
                             text=volcano_labs[[cont]][[status]],
                             hoverinfo='text',
                             mode = 'markers',
                             marker = list(color=status_colors[[status]]),
                             visible = TRUE,
                             name = all_statuses[[status]])
  }
}

n_status_keys <- length(all_status_keys)
n_contrasts <- length(volcano_contrasts)
start_index <- 1
all_buttons <- list()

# ideally this would be done with lapply but this will work
for (item in c(1:n_contrasts)){
  cur_vec <- rep(FALSE, n_contrasts*n_status_keys)
  cur_vec[start_index:(start_index+n_status_keys-1)] <- TRUE
  start_index <- start_index + n_status_keys
  cur_list <- as.list(c(FALSE, cur_vec))
  cur_button <- list(
    method = "restyle",
    args  = list("visible", cur_list),
    label = volcano_contrasts[item])
  all_buttons <- append(all_buttons, list(cur_button))
}

updatemenus <- list(
  list(
    active = 0,
    buttons = all_buttons
  )
)

fig <- fig %>% layout(
                      showlegend = TRUE,
                      updatemenus = updatemenus,
                      xaxis = list(title = 'log2 Fold Change'),
                      yaxis = list(title = '-log10 adjusted p-value')
                      )
fig
```


# PCA plots {.tabset .tabset-fade}

```{r PCA_function, collapse=TRUE, out.width="10px", out.height="10px"}


plot_interactive_PCA <- function(rld_data, intgroup_to_plot, all_intgroups){
  ## Perform PCA analysis
  data_pca <- plotPCA(rld_data, intgroup = all_intgroups, ntop = nrow(SummarizedExperiment::assay(rld_data)), returnData = TRUE)
  ## Get percent of variance explained
  percentVar <- round(100 * attr(data_pca, "percentVar"))
  ## Plot
  highlight_pca <- highlight_key(data_pca, ~group) # is this going to fail if the data doesn't have a "group" variable?
  # colour stuff
  colour_scheme <- "Set1"
  if (length(unique(SummarizedExperiment::colData(rld_data)[, intgroup_to_plot])) > 9 ) { # max for Set1
    colour_scheme <- "Spectral" # not my favourite but this plot isn't going to work for large numbers of colours anyway.
  }
  p <- plot_ly(data = highlight_pca,
               x = ~PC1,
               y = ~PC2,
               color = as.formula(paste0("~",intgroup_to_plot)),
               text = ~name,
               type = "scatter",
               mode = "markers",
               colors = colour_scheme) %>%
    layout(xaxis = list(title = paste0("PC1: ", percentVar[1] ,"% variance")),
           yaxis = list(title = paste0("PC2: ", percentVar[2] ,"% variance"))
    )
  plotly::highlight(p, on = "plotly_click", off = "plotly_doubleclick", color = "red")

  
  return(p)
}

if(params$platform == "RNA-Seq"){
  plot_ly() %>% plotly::layout(autosize = F, width = 10, height = 10)
  # need this initialization for the following plots to show up
  # since plotly isn't properly initialized in a loop and the volcano above isn't run for RNA-seq
}

```

This plot shows the first two principal components that explain the variability in the data using the regularized log count data. If you are unfamiliar with principal component analysis, you might want to check the [Wikipedia entry](https://en.wikipedia.org/wiki/Principal_component_analysis) or this [interactive explanation](http://setosa.io/ev/principal-component-analysis/).


## All `r nrow(rld)` genes {.tabset .tabset-fade}

```{r PCA_all_genes, results='asis', warning = FALSE}
for (g in params$intgroup_to_plot){
  cat(paste0("### ",g,"\n\n"))
  print(htmltools::tagList( plot_interactive_PCA(rld, g, params$intgroup_to_plot)))
  cat("\n\n")
}
```

## `r nrow(rld_DEGs)` DEGs {.tabset .tabset-fade}

```{r PCA_DEGs, results='asis', warning = FALSE}
if (length(rld_DEGs) > 1) {
  for (g in params$intgroup_to_plot){
    cat(paste0("### ",g,"\n\n"))
    print(htmltools::tagList( plot_interactive_PCA(rld_DEGs, g, params$intgroup_to_plot)))
    cat("\n\n")
  }
} else {
  print("Insufficient DEGs found.")
}
```

## Top `r params$nBest` genes {.tabset .tabset-fade}

```{r PCA_top_N, results='asis', warning = FALSE}
for (g in params$intgroup_to_plot){
  cat(paste0("### ",g,"\n\n"))
  print(htmltools::tagList( plot_interactive_PCA(rld_top,g, params$intgroup_to_plot)))
  cat("\n\n")
} 
```

# DEG heatmaps {.tabset .tabset-fade}
```{R heatmap_function}
plot_DEG_heatmap <- function(rld_data){
  mat <- SummarizedExperiment::assay(rld_data) %>%
  as_tibble(rownames="Feature_ID") %>%
  left_join(id_table, by="Feature_ID") %>%
  mutate(Gene_Symbol_2 = coalesce(Gene_Symbol, Feature_ID)) %>%
  mutate(label = paste0(Gene_Symbol_2," : ",Feature_ID)) %>%
  dplyr::select(-Ensembl_Gene_ID, -Gene_Symbol, -Gene_Symbol_2, -Feature_ID) %>%
  tibble::column_to_rownames("label")
  if (nrow(as.matrix(mat)) > 1) {
    mat <- mat - rowMeans(mat)
    return(pheatmap(mat,
             annotation_col =  as.data.frame(SummarizedExperiment::colData(rld_DEGs)[, params$intgroup_to_plot, drop = F]),
             show_rownames = (nrow(rld_data) <= 50),
             border_color = NA,
             scale = "row"))
  } else {
    return("No DEGs found.")
  }
}


```


## All DEGs
```{r DEG_heatmap_all_genes,fig.width = 10, fig.height = 10}
plot_DEG_heatmap(rld_DEGs)
```

## Top `r params$nHeatmapDEGs` differentially abundant genes
```{r DEG_heatmap_DEGs, fig.width = 10, fig.height = 10}
plot_DEG_heatmap(rld_top_heatmap)
```

# Genes of interest

This section contains plots showing the normalized counts per sample for each group of interest. Only the best `r params$nBestFeatures` features (adjusted p-value <`r params$alpha`) are shown, ranked by their absolute fold change values.

## Overview
```{r genes_of_interest, fig.width = 10, fig.height = 10}

numResults <- params$nBestFeatures

allResultsOrdered_logFC_filter <- significantResults %>% arrange(-abs(linearFoldChange))


allResultsOrdered_logFC_filter %>%
  group_by(contrast) %>%
  top_n(numResults, wt = abs(log2FoldChange)) %>%
  ungroup() %>%
  mutate(Contrast = as.factor(contrast),
         Symbol = reorder_within(Gene_Symbol, log2FoldChange, Contrast)) %>%
  ggplot(aes(x = log2FoldChange,
             y = Symbol,
             color = Contrast,
             size = -log(padj))) +
  geom_point(show.legend = TRUE) +
  facet_wrap(~contrast,
             scales = "free_y",
             ncol = 4,
             labeller = labeller(contrast = label_wrap_gen(10))) +
  scale_y_reordered()  +
  geom_vline(xintercept = 0,
             linetype = "dashed",
             color = "black",
             size = 1) +
  theme_bw() +
  ggtitle(paste0("Top ",numResults," genes by fold change grouped by treatment"))

```

## By gene {.tabset .tabset-fade}

The Y axis is on the log10 scale and the feature name is shown in the title of each plot.

```{r genelevel_plots, results='asis', message = FALSE, fig.width = 5, fig.height = 5}

nBestFeatures <- params$nBestFeatures

# Would be valuable to have this broken down by contrast too.
plotCounts_gg <- function(i, d, intgroup) {
    plotdata <- plotCounts(d,
                       gene = i,
                       intgroup = intgroup,
                       returnData = TRUE)
    plot_title <- paste(id_table[["Gene_Symbol"]][grep(i,id_table[["Feature_ID"]])],
                        id_table[["Ensembl_Gene_ID"]][grep(i, id_table[["Feature_ID"]])])
    if(params$platform == "TempOSeq"){
      plot_title <- paste(plot_title, id_table[["Feature_ID"]][grep(i, id_table[["Feature_ID"]])])
    }
    if (ncol(plotdata) > 2) {
      colorCol = 3
    } else {colorCol = 2}
    ggplot(plotdata, aes(x = plotdata[, 2], y = plotdata[, 1], color = plotdata[, colorCol])) + 
      geom_point(position = position_jitterdodge()) +
      ylab('Normalized count') +
      xlab('Group') +
      ggtitle(plot_title) +
      coord_trans(y = "log10") +
      theme_bw() +
      theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
      labs(color = colnames(plotdata)[colorCol])
}

genesToPlot <- significantResults %>% arrange(-abs(log2FoldChange))
for (i in head(unique(genesToPlot[["Feature_ID"]]), nBestFeatures)) {
  cat("\n###", id_table[["Gene_Symbol"]][grep(i,id_table[["Feature_ID"]])], "  \n\n")
  print(plotCounts_gg(i, d = dds, intgroup = params$design))
  cat('\n\n')
}


```


# Methods Summary

Please use this as a starting point for the bioinformatics/statistics section of any publications based on the data analyzed in this report. Note that there is a similar paragraph at the end of the sequencing quality control report.


  A count matrix containing the $`r paste(sample_count_metadata$samples_postQC)`$ samples that passed QC was imported into R for statistical analysis. After removing samples with less than $`r params$nmr_threshold`$ reads, `r sample_count_metadata$samples_filtered` samples were left. Following the recommendations set out by the Omics Data Analysis Frameworks for Regulatory application (R-ODAF) guidelines [@VERHEIJEN2022105143], genes were filtered to include only those where 75% of at least one experimental group were above `r params$MinCount` CPM, and spurious spikes were removed in which (max - median) of counts were less than (sum of counts)/(number of replicates + 1). We used DESeq2 `r packageVersion("DESeq2")` [@ggplotcitation] to test for differentially abundant genes within the `r params$platform` data. The log2FoldChange shrinkage procedure used was `r resultsListAll[[1]]@priorInfo$type`. An alpha of `r resultsListAll[[1]]@metadata$alpha` was used to extract raw results, which are reported as the `r gsub(mcols(x=resultsListAll[[1]])$description[[4]],pattern=":.*",replacement="")`. To account for multiple testing, `r mcols(resultsListAll[[1]])$description[5]` are reported. Cook's cutoff was set to `r params$cooks` in this analysis (a value of FALSE indicates that Cook's distance filtering was disabled). Differentially expressed genes (DEGs) were filtered using a linear fold change cutoff of `r params$linear_fc_filter_DEGs` and adjusted p-value of `r params$alpha`.


## References {-}

If these methods were used in your study, please cite the following papers as appropriate:  

---
nocite: '@*'
---

<div id="refs"></div>


```{r disconnect, include = F}
DBI::dbDisconnect(dbconn(db))
```

```{r, child=file.path(paths$root,'Rmd','session_info.Rmd')}
```