---
params:
  projectdir: null # Use when loading. If null, here::here() will be used. Can also hard code if for some reason you need that
  project_title: "Test"    # Change depending on your project name
  project_name: "Name" 
  project_description: null # description of project for report, optional
  batch_var: "sample_plate"          # For example, multiple plates
  dose: "dose"               # If there is a dose in the experiment; otherwise use NULL
  platform: "TempO-Seq"      # TempO-Seq Or RNA-Seq
  nmr_threshold: 100000    # 10% of 1M reads for TempOSeq = 100,000; 10% of 10M reads for RNA-Seq = 1,000,000.
  write_additional_output: FALSE # export BMD, biomarker, and biosets output? facet variable should contain chemical and timepoint information
  celltype: "Some cells" # only required for biosets output. Cell type or species used
  units: "uM" # only required for biosets output. Units of dose
  species: "human"                 # one of human, mouse, rat, hamster
  design: "group"                # single experimental group of interest; entries in this column must match the contrast names.
  intgroup: ["group"]            # experimental group of interest plus covariates; can be more than one
  intgroup_to_plot: ["group"]   # for PCA plots, add tabs for one or more groups to color
  formula_override: null # e.g., "~batch + condition", to enable users to specify the DESeq2 formula to use
  group_facet: "chemical"       # If you have many different experimental groups, you may subset the report by specifying a column in the metadata to filter groups, and then setting the group of interest in group_filter
  group_filter:  null        # Which group will this report be done on?
  display_group_facet: null #  
  display_group_filter:  null # 
  sortcol: "dose"                # Optionally, a column by which to sort the contrasts
  lenient_contrasts: FALSE         # Use either column (exp, cont) of contrasts file to limit what is included in the report (instead of just exp column)
  strict_contrasts: FALSE         # Use BOTH columns (exp, cont) of contrasts file to limit what is included in the report
  exclude_samples: null      # Optionally, a vector of sample names to exclude from the analysis
  exclude_groups: null       # Optionally, a vector of groups to exclude from the analysis. By default this is assumed to be in the column specified by params$design.
  include_only_column:  null # Restrict analysis to group(s) in the column listed here based on params$include_only_group.
  include_only_group:  null  # Restrict analysis to this/these group(s) within the column listed in params$include_only_column
  cpus: 41                       # Set to a lower number (e.g., 2 to 4) if you aren't working in a server environment
  run_pathway_analysis: TRUE     # Optionally disable pathway analysis if not available for your organism
  wikipathways_directory: "~/shared/dbs/wikipathways"
  biospyder_dbs: "./unify_temposeq_manifests/output_manifests"
  biospyder_manifest_file:  "Human_S1500_1.2_standardized.csv"
  # "191113_Human_S1500_Surrogate_2.0_Manifest.csv"
  # "181019_Human_S1500_Surrogate_1.2_Manifest.txt"
  # "191004_Human_Whole_Transcriptome_2.0_Manifest.txt"
  wikipathways_filename: "wikipathways-20210810-gmt-Homo_sapiens.gmt"
  # "wikipathways-20210810-gmt-Mus_musculus.gmt"
  # "wikipathways-20210810-gmt-Rattus_norvegicus.gmt"
  nBestFeatures: 20              # The number of best features to make plots of their counts
  nBest: 100                     # Number of features to include in table and limiting PCA/clustering analysis
  nHeatmap: 50                   # Number of most variable genes for heatmap
  nHeatmapDEGs: 50               # Number of DEGs for heatmap
  cooks: FALSE                   # the DESeq Cook's distance cutoff, or FALSE to disable it
  filter_gene_counts: FALSE      # Filter genes to those with at least 1 count in 1 sample. Disabled by default for biomarker file output. Can enable to increase speed of analysis
  generate_main_report: TRUE
  generate_stats_report: TRUE
  generate_data_explorer_report: TRUE
  generate_go_pathway_report: TRUE
  output_digits: 5
  parallel: FALSE
  species_data: null
  count_data_file: null
  sampledata_sep: null
  MinCount: null
  alpha: null
  linear_fc_filter: null
  feature_id: null
  biospyder: null
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    code_folding: hide
    theme: spacelab           # flatly spacelab sandstone cerulean
    code_download: true
---


```{r docSetup, warning = FALSE, message = FALSE}
#### Record start time
startTime <- Sys.time()

library('clusterProfiler')
library('enrichplot')

source(here::here("scripts","setup_functions.R"), local = TRUE)
source(here::here("scripts","DESeq_functions.R"), local = TRUE)
source(here::here("scripts","data_functions.R"), local = TRUE)

species_data <- load_species(params$species, params$wikipathways_filename, params$biospyder_manifest_file)
db <- AnnotationDbi::loadDb(dbfile(get(species_data$orgdb)))

# now that we have all our params etc loaded, we can set the HTML title of our document 
```
---
title: "`r params$platform` DEG analysis - GO and pathway analysis"
---



# `r params$project_title` - GO and pathway analysis {-}

Date report generated: `r format(Sys.time(), '%d %B, %Y')`  

Report prepared by: `r params$project_bioinformatics_name`

Report prepared for: `r params$project_name`


***

This report contains Gene Ontology and pathway analysis.

```{r load_facet_data, include=FALSE}
source(here::here("scripts","load-facet-data.R"), local = TRUE)
```



# GO Enrichment Analysis {.tabset .tabset-fade}

This section performs GO enrichment on **all DEGs passing filters**. The background set of genes is those that were identified in this sequencing experiment. The [clusterProfiler](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) package is used for this analysis.

```{r cluster-profiler, warning = FALSE, message = FALSE, collapse = TRUE}

if(params$platform == "TempO-Seq"){
id_table_entrez <- params$biospyder %>%
  dplyr::select(Feature_ID=Probe_Name, Gene_Symbol, Entrez_ID)
} else {
  id_table_entrez <- AnnotationDbi::select(db,
                                           columns = c("ENSEMBL", "SYMBOL", "ENTREZID"),
                                           keys = allResults$Ensembl_Gene_ID, keytype="ENSEMBL") %>%
    distinct()
  colnames(id_table_entrez) <- c("Feature_ID","Gene_Symbol","Entrez_ID")
  id_table_entrez$Entrez_ID <- as.integer(id_table_entrez$Entrez_ID)
}

probeIDDEGs <- data.frame(genes=significantResults[["Feature_ID"]])
names(probeIDDEGs)[names(probeIDDEGs) == 'genes'] <- "Feature_ID"

DEGs <- dplyr::left_join(probeIDDEGs, id_table_entrez, by = "Feature_ID")
entrezDEGs <- as.character(DEGs$Entrez_ID)
entrezDEGs <- entrezDEGs[!is.na(entrezDEGs)] %>% unique()

background <- dplyr::left_join(data.frame(genes = row.names(assay(dds))),
                               id_table_entrez,
                               by = c("genes" = "Feature_ID")) %>%
  dplyr::filter(!is.na(Entrez_ID)) %>%
  dplyr::pull(Entrez_ID)
background <- as.character(background)

KeyType <- "ENTREZID"
```

```{r enrich_go, warning = FALSE, message = FALSE, collapse = TRUE}
enrich_go_bp <- enrichGO(gene = entrezDEGs,
                         universe = background,
                         OrgDb = db,
                         keyType = KeyType,
                         readable = T,
                         ont = "BP",
                         pAdjustMethod = "BH",
                         pvalueCutoff = 0.01,
                         qvalueCutoff = 0.05)


enrich_go_mf <- enrichGO(gene = entrezDEGs,
                         universe = background,
                         OrgDb = db,
                         keyType = KeyType,
                         readable = T,
                         ont = "MF",
                         pAdjustMethod = "BH",
                         pvalueCutoff = 0.01,
                         qvalueCutoff = 0.05)

enrich_go_cc <- enrichGO(gene = entrezDEGs,
                         universe = background,
                         OrgDb = db,
                         keyType = KeyType,
                         ont = "CC",
                         pAdjustMethod = "BH",
                         pvalueCutoff = 0.01,
                         qvalueCutoff = 0.05)


# TODO GO table output?
# write.table(as.data.frame(enrich_go_bp),
#             file = file.path(paths$pathway_analysis, "Enriched.GO_terms.Biological_Processes.txt"))
# write.table(as.data.frame(enrich_go_mf),
#             file = file.path(paths$pathway_analysis, "Enriched.GO_terms.Molecular_Functions.txt"))
# write.table(as.data.frame(enrich_go_cc),
#             file = file.path(paths$pathway_analysis, "Enriched.GO_terms.Cellular_Component.txt"))


# get fold-changes

DEGs_full <- dplyr::left_join(significantResults,
                              id_table_entrez,
                              by = "Feature_ID") %>%
  distinct()
foldChanges <- DEGs_full %>% dplyr::pull(log2FoldChange)
names(foldChanges) <- DEGs_full %>% dplyr::pull(Entrez_ID)
foldChanges <- foldChanges %>% sort() %>% rev()
foldChanges <- foldChanges[!is.na(names(foldChanges))]


```

## Biological processes


```{r plot_GO_analysis_bp, warning = FALSE, message = FALSE, fig.width = 12, collapse = TRUE}
if (!is.null(enrich_go_bp)) {
  if (nrow(enrich_go_bp@result %>% filter(p.adjust < 0.01 & qvalue < 0.05)) > 3) {
    plot1 <- clusterProfiler::dotplot(enrich_go_bp, font.size = 9, showCategory = 20)  +
      theme(axis.text.y = element_text(angle = 0)) +
      scale_y_discrete(labels = function(x) stringr::str_wrap(x, width = 30))

    simplified_bp <- simplify(enrich_go_bp, cutoff = 0.7, by = "p.adjust", select_fun = min)
    # Max size of gene set... useful for visualizing networks
    simplified_bp_for_network <- simplified_bp %>% DOSE::gsfilter(by = "GSSize", max = 400)
    # I believe there is a bug in the current version of DOSE or clusterProfiler that prevents this from working
    # Throws this error on plotting: Error in graph.data.frame(x, directed = FALSE) : the data frame should contain at least two columns
    simplified_bp_filt <- simplified_bp %>%
      filter(p.adjust < .05, qvalue < 0.05) %>%
      mutate(GeneRatio = DOSE::parse_ratio(GeneRatio)) %>%
      arrange(desc(GeneRatio))

    if (nrow(as.data.frame(simplified_bp_filt)) < 10) {
        show_num <- simplified_bp_filt$Description # Top N less than 10
      } else {
        show_num <- simplified_bp_filt$Description[1:10] # Top 10
      }

    plot2 <- cnetplot(simplified_bp,
                      foldChange = foldChanges,
                      showCategory = show_num,
                      cex_label_gene = 0.3)
    
    plot3 <- upsetplot(enrich_go_bp)
    plot4 <- heatplot(simplified_bp, foldChange = foldChanges, showCategory = show_num) +
      scale_y_discrete(labels = function(x) stringr::str_trunc(x, width = 50))
    
    print(plot1)
    print(plot2)
    print(plot3)
    print(plot4)
  } else {
    num_terms <- nrow(enrich_go_bp@result %>% filter(p.adjust < 0.01 & qvalue < 0.05))
    cat(paste0(num_terms," enriched terms found. This is not enough to report in the results."))
  }
} else { cat("No significantly enriched terms using criteria selected") }
```


## Molecular Functions

```{r plot_GO_analysis_mf, warning = FALSE, message = FALSE, fig.width = 12, collapse = TRUE}
if (!is.null(enrich_go_mf)) {
  
  if (nrow(enrich_go_mf@result %>% filter(p.adjust < 0.01 & qvalue < 0.05)) > 3) {
    plot1 <- clusterProfiler::dotplot(enrich_go_mf, font.size = 9)  +
      theme(axis.text.y = element_text(angle = 0)) +
      scale_y_discrete(labels = function(x) stringr::str_wrap(x, width = 30))
    
    simplified_mf <- simplify(enrich_go_mf, cutoff = 0.7, by = "p.adjust", select_fun = min)
    # Max size of gene set... useful for visualizing networks
    simplified_mf_for_network <- simplified_mf %>% DOSE::gsfilter(by = "GSSize", max = 400)
    # I believe there is a bug in the current version of DOSE or clusterProfiler that prevents this from working
    # Throws this error on plotting: Error in graph.data.frame(x, directed = FALSE) : the data frame should contain at least two columns
    simplified_mf_filt <- simplified_mf %>%
      filter(p.adjust < .05, qvalue < 0.05) %>%
      mutate(GeneRatio = DOSE::parse_ratio(GeneRatio)) %>%
      arrange(desc(GeneRatio))
    
    if (nrow(as.data.frame(simplified_mf)) < 10) {
        show_num <- simplified_mf_filt$Description # Top N less than 10
      } else {
        show_num <- simplified_mf_filt$Description[1:10] # Top 10
      }
    
    if (nrow(simplified_mf) > 3) {
      plot2 <- cnetplot(simplified_mf,
                        foldChange = foldChanges,
                        showCategory = show_num,
                        cex_label_gene = 0.3)
    }
    
    enrich_go_mf_upset <- enrich_go_mf
    enrich_go_mf_upset@result$Description <- str_trunc(enrich_go_mf_upset@result$Description, width = 50)
    plot3 <- upsetplot(enrich_go_mf_upset)
    plot4 <- enrichplot::heatplot(simplified_mf, foldChange = foldChanges, showCategory = show_num) +
      scale_y_discrete(labels = function(x) stringr::str_trunc(x, width = 50))
    
    print(plot1)
    if (nrow(simplified_mf) > 3) {print(plot2)} else {print("Network plot not shown due to a limited number of enriched terms.")}
    print(plot3)
    print(plot4)
  } else {
    num_terms <- nrow(enrich_go_mf@result %>% filter(p.adjust < 0.01 & qvalue < 0.05))
    cat(paste0(num_terms," enriched terms found. This is not enough to report in the results."))
  }
} else {cat("No significantly enriched terms using criteria selected") }
```

## Cellular Component

```{r plot_GO_analysis_cc, warning = FALSE, message = FALSE, fig.width = 12, collapse = TRUE}
if (!is.null(enrich_go_cc)) {
  if (nrow(enrich_go_cc@result %>% filter(p.adjust < 0.01 & qvalue < 0.05)) > 3) {
    plot1 <- clusterProfiler::dotplot(enrich_go_cc, font.size = 9)  +
      theme(axis.text.y = element_text(angle = 0)) +
      scale_y_discrete(labels = function(x) stringr::str_wrap(x, width = 30))
    
    simplified_cc <- simplify(enrich_go_cc, cutoff = 0.7, by = "p.adjust", select_fun = min)
    # Max size of gene set... useful for visualizing networks
    simplified_cc_for_network <- simplified_cc %>% DOSE::gsfilter(by = "GSSize", max = 400)
    # I believe there is a bug in the current version of DOSE or clusterProfiler that prevents this from working
    # Throws this error on plotting: Error in graph.data.frame(x, directed = FALSE) : the data frame should contain at least two columns
    simplified_cc_filt <- simplified_cc %>%
      filter(p.adjust < .05, qvalue < 0.05) %>%
      mutate(GeneRatio = DOSE::parse_ratio(GeneRatio)) %>%
      arrange(desc(GeneRatio))

    if (nrow(as.data.frame(simplified_cc_filt)) < 10) {
      show_num <- simplified_cc_filt$Description # Top N less than 10
      } else {
        show_num <- simplified_cc_filt$Description[1:10] # Top 10
      }
    
    plot2 <- cnetplot(simplified_cc,
                      foldChange = foldChanges,
                      showCategory = show_num,
                      cex_label_gene = 0.3)
    
    enrich_go_cc_upset <- enrich_go_cc
    enrich_go_cc_upset@result$Description <- str_trunc(enrich_go_cc_upset@result$Description, width = 50)
    plot3 <- upsetplot(enrich_go_cc_upset)
    plot4 <- enrichplot::heatplot(simplified_cc, foldChange = foldChanges, showCategory = show_num) +
      scale_y_discrete(labels = function(x) stringr::str_trunc(x, width = 50))
    
    print(plot1)
    print(plot2)
    print(plot3)
    print(plot4)
  } else {
    num_terms <- nrow(enrich_go_cc@result %>% filter(p.adjust < 0.01 & qvalue < 0.05))
    cat(paste0(num_terms," enriched terms found. This is not enough to report in the results."))
  }
} else {cat("No significantly enriched terms using criteria selected")}
```


# Pathway Analysis using Enriched WikiPathways and GSEA (KEGG)

## WikiPathways: Enrichment based on all DEGs

This section uses clusterProfiler to detect enriched gene sets using [WikiPathways](https://www.wikipathways.org/) data.  

```{r 'pathway_analysis', collapse=TRUE, eval=params$run_pathway_analysis, include=params$run_pathway_analysis, warning = FALSE, message = FALSE}
entrez_IDs <- as.character(id_table_entrez$Entrez_ID)
entrezDEGs <- dplyr::left_join(significantResults,
                               id_table_entrez,
                               by = "Feature_ID")
entrezDEGs <- as.vector(entrezDEGs$Entrez_ID)

#Wiki <- rWikiPathways::downloadPathwayArchive(organism=species_sci, format = "gmt", destpath = paths$DEG_output, date="20210310")
wp2gene <- clusterProfiler::read.gmt(file.path(paths$wikipathways,params$wikipathways_filename))
wp2gene <- wp2gene %>% tidyr::separate(1, c("name","version","wpid","org"), "%")
wpid2gene <- wp2gene %>% dplyr::select(wpid,gene) #TERM2GENE
wpid2name <- wp2gene %>% dplyr::select(wpid,name) #TERM2NAME
wpid2gene
wpid2name
matches <- wpid2gene %>% filter(gene %in% entrezDEGs)

if(is.null(params$display_group_filter)){
  pathway_output_dir <- paths$pathway_analysis
} else {
  pathway_output_dir <- paths$pathway_analysis[[params$display_group_filter]]
}

if (nrow(matches) > 1) {
  DEG_pathways <- clusterProfiler::enricher(
    entrezDEGs,
    universe = entrez_IDs,
    pAdjustMethod = "fdr",
    pvalueCutoff = 0.05, #p.adjust cutoff
    TERM2GENE = wpid2gene,
    TERM2NAME = wpid2name)
  DEG_pathways <- DOSE::setReadable(DEG_pathways, db, keyType = "ENTREZID")
  if (nrow(DEG_pathways@result %>% filter(p.adjust < 0.01 & qvalue < 0.05)) > 1) {
    write.table(as.data.frame(DEG_pathways),
                file = file.path(pathway_output_dir,"Enriched.WikiPathways.txt"))
  } else { cat("No significantly enriched terms using criteria selected") }
}
```


## WikiPathways: Enrichment by DEGs within each contrast {.tabset .tabset-fade}

This section uses clusterProfiler `enricher()` to detect enriched gene sets using [WikiPathways](https://www.wikipathways.org/) data. Only contrasts with results are shown here.  

```{r 'pathway_analysis_by_contrast_make_plots', collapse=TRUE, eval=params$run_pathway_analysis, include=params$run_pathway_analysis, warning = FALSE, message = FALSE}
plotList <- list()
contrasts <- levels(significantResults$contrast)
for (i in seq_along(contrasts) ){
  entrezDEGs <- dplyr::left_join(significantResults %>% filter(contrast == contrasts[i]),
                                 id_table_entrez,
                                 by = "Feature_ID")
  entrezDEGs <- entrezDEGs$Entrez_ID
  if (length(entrezDEGs) < 10) {
    plotList[[i]] <- NA
    next
  }
  DEG_pathways <- clusterProfiler::enricher(entrezDEGs,
                                            universe = entrez_IDs,
                                            pAdjustMethod = "fdr",
                                            pvalueCutoff = 0.05, #p.adjust cutoff
                                            TERM2GENE = wpid2gene,
                                            TERM2NAME = wpid2name)
  DEG_pathways <- DOSE::setReadable(DEG_pathways, db, keyType = "ENTREZID")

  if (nrow(DEG_pathways@result %>% filter(p.adjust < 0.01 & qvalue < 0.05)) > 1) {
    plotList[[i]] <- clusterProfiler::dotplot(DEG_pathways, showCategory = 20, font.size = 9)  +
      theme(axis.text.y = element_text(angle = 0)) +
      scale_y_discrete(labels = function(x) stringr::str_wrap(x, width = 30))
    write.table(as.data.frame(DEG_pathways),
                file = file.path(pathway_output_dir,paste0("Enriched.WikiPathways.",
                                                         str_replace_all(contrasts[i]," ","_"),
                                                         ".txt")))
  } else { plotList[[i]] <- NA }
}
```

```{r 'pathway_analysis_by_contrast_print_plots', results='asis', eval=params$run_pathway_analysis, include=params$run_pathway_analysis, warning = FALSE, message = FALSE}
for (i in seq_along(contrasts)) {
  if (any(!is.na(plotList[[i]]))) {
    cat("###", contrasts[i], "  \n\n")
    print(plotList[[i]])
    cat('  \n\n')
  }
}

```


## GSEA: By contrast {.tabset .tabset-fade}

This section uses clusterProfiler `gseKEGG()` to detect enriched KEGG pathways and enrichplot `gseaplot2` to plot running scores.

```{r 'gsea_analysis_by_contrast_make_plots', collapse=TRUE, eval=params$run_pathway_analysis, include=params$run_pathway_analysis,  warning = FALSE, message = FALSE}

plotListGSEA <- list()
plotListGSEA_escore <- list()
contrasts <- levels(allResults$contrast)
bg_genes <-read.gmt("~/storage/LearnPipline/R-ODAF_Health_Canada/c2.cp.kegg_medicus.v2023.2.Hs.symbols.gmt") 

for (i in seq_along(contrasts)) {
  DEGs_full <- dplyr::left_join(allResults %>% filter(contrast ==  contrasts[i]),
                                id_table_entrez,
                                by = "Feature_ID")
  foldChanges <- DEGs_full %>% dplyr::pull(log2FoldChange)
  names(foldChanges) <- DEGs_full %>% dplyr::pull(Gene_Symbol.x)
  foldChanges <- foldChanges %>% sort() %>% rev()
  foldChanges <- foldChanges[!is.na(names(foldChanges))]
  if (length(foldChanges) < 10) {
    plotListGSEA[[i]] <- NA
    plotListGSEA_escore[[i]] <- NA
    next
  }
  kk <- GSEA(foldChanges, pvalueCutoff = 0.05, TERM2GENE = bg_genes,)
  gsea_sorted <- kk %>% arrange(desc(enrichmentScore))
  gsea_sig <- gsea_sorted %>% filter(p.adjust < 0.05)
  write.table(as.data.frame(gsea_sorted),
              file = file.path(pathway_output_dir,paste0("Enriched.KEGG_Pathways.",
                                                       str_replace_all(contrasts[i]," ","_"),
                                                       ".txt")))
  if (dim(kk@result)[1] > 1) {
    plotListGSEA[[i]] <- ridgeplot(kk)
    plotListGSEA_escore[[i]] <- enrichplot::gseaplot2(gsea_sig, geneSetID = c(seq_along(gsea_sig$ID)))
  } else {
    plotListGSEA[[i]] <- NA
    plotListGSEA_escore[[i]] <- NA
  }
}
```

```{r 'gsea_analysis_by_contrast_print_plots', results='asis', eval=params$run_pathway_analysis, include=params$run_pathway_analysis,  warning = FALSE, message = FALSE}
for (i in seq_along(contrasts)) {
  if (any(!is.na(plotListGSEA[[i]]))) {
    cat("###", contrasts[i], "  \n\n")
    print(plotListGSEA[[i]])
    cat('  \n\n')
  }
}
```

## GSEA Enrichment Score Plots: By contrast {.tabset .tabset-fade}

In this section, only the pathway with the top enrichment score is shown (for simplicity). Note that it is possible to plot any pathway of interest by altering the call to `gseaplot`.

```{r 'gsea_scores_plot', results='asis', eval=params$run_pathway_analysis, include=params$run_pathway_analysis,  warning = FALSE, message = FALSE}
for (i in seq_along(contrasts)) {
  if (any(!is.na(plotListGSEA_escore[[i]]))) {
    cat("###", contrasts[i], "  \n\n")
    print(plotListGSEA_escore[[i]])
    cat('  \n\n')
  }
}
```

```{r disconnect, include = F}
DBI::dbDisconnect(dbconn(db))
```

```{r, child=file.path(paths$root,'Rmd','session_info.Rmd')}
```