---
params:
  projectdir: null # Use when loading. If null, here::here() will be used. Can also hard code if for some reason you need that
  project_title: "Liver toxic chemicals"    # Change depending on your project name
  project_name: "Your name here" 
  project_description: null # description of project for report, optional
  batch_var: "batch"          # For example, multiple plates
  dose: "dose"               # If there is a dose in the experiment; otherwise use NULL
  platform: "TempO-Seq"      # TempO-Seq Or RNA-Seq
  nmr_threshold: 100000    # 10% of 1M reads for TempOSeq = 100,000; 10% of 10M reads for RNA-Seq = 1,000,000.
  write_additional_output: TRUE # export BMD, biomarker, and biosets output? facet variable should contain chemical and timepoint information
  celltype: "Liver spheroids" # only required for biosets output. Cell type or species used
  units: "uM" # only required for biosets output. Units of dose
  species: "human"                 # one of human, mouse, rat, hamster
  design: "Group"                # single experimental group of interest; entries in this column must match the contrast names.
  intgroup: ["Group"]            # experimental group of interest plus covariates; can be more than one
  intgroup_to_plot: ["group"]   # for PCA plots, add tabs for one or more groups to color
  formula_override: null # e.g., "~batch + condition", to enable users to specify the DESeq2 formula to use
  group_facet: "Chemical_S9"       # If you have many different experimental groups, you may subset the report by specifying a column in the metadata to filter groups, and then setting the group of interest in group_filter
  group_filter:  null        # Which group will this report be done on?
  display_group_facet: "chemical_timepoint" #  
  display_group_filter:  "estradiol 10d" # 
  sortcol: "dose"                # Optionally, a column by which to sort the contrasts
  lenient_contrasts: FALSE         # Use either column (exp, cont) of contrasts file to limit what is included in the report (instead of just exp column)
  strict_contrasts: FALSE         # Use BOTH columns (exp, cont) of contrasts file to limit what is included in the report
  exclude_samples: null      # Optionally, a vector of sample names to exclude from the analysis
  exclude_groups: null       # Optionally, a vector of groups to exclude from the analysis. By default this is assumed to be in the column specified by params$design.
  include_only_column:  null # Restrict analysis to group(s) in the column listed here based on params$include_only_group.
  include_only_group:  null  # Restrict analysis to this/these group(s) within the column listed in params$include_only_column
  cpus: 41                       # Set to a lower number (e.g., 2 to 4) if you aren't working in a server environment
  run_pathway_analysis: TRUE     # Optionally disable pathway analysis if not available for your organism
  wikipathways_directory: "~/shared/dbs/wikipathways"
  biospyder_dbs: "~/shared/dbs/biospyder/"
  biospyder_manifest_file:  "Human_S1500_Surrogate_2.0_standardized.csv"
  # "191113_Human_S1500_Surrogate_2.0_Manifest.csv"
  # "181019_Human_S1500_Surrogate_1.2_Manifest.txt"
  # "191004_Human_Whole_Transcriptome_2.0_Manifest.txt"
  wikipathways_filename: "wikipathways-20210810-gmt-Homo_sapiens.gmt"
  # "wikipathways-20210810-gmt-Mus_musculus.gmt"
  # "wikipathways-20210810-gmt-Rattus_norvegicus.gmt"
  nBestFeatures: 20              # The number of best features to make plots of their counts
  nBest: 100                     # Number of features to include in table and limiting PCA/clustering analysis
  nHeatmap: 50                   # Number of most variable genes for heatmap
  nHeatmapDEGs: 50               # Number of DEGs for heatmap
  cooks: FALSE                   # the DESeq Cook's distance cutoff, or FALSE to disable it
  filter_gene_counts: FALSE      # Filter genes to those with at least 1 count in 1 sample. Disabled by default for biomarker file output. Can enable to increase speed of analysis
  generate_main_report: TRUE
  generate_stats_report: TRUE
  generate_data_explorer_report: TRUE
  generate_go_pathway_report: TRUE
  output_digits: 5
  parallel: FALSE
  species_data: null
  count_data_file: null
  sampledata_sep: null
  MinCount: null
  alpha: null
  linear_fc_filter: null
  feature_id: null
  biospyder: null
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    code_folding: hide
    theme: spacelab           # flatly spacelab sandstone cerulean
    code_download: true
---

```{r docSetup, warning = FALSE, message = FALSE}
#### Record start time
startTime <- Sys.time()

library('flextable')
library('vsn')
library('RColorBrewer')
library('pheatmap')
source(here::here("scripts","DESeq_functions.R"), local = TRUE)
source(here::here("scripts","data_functions.R"), local = TRUE)

species_data <- load_species(params$species, params$wikipathways_filename, params$biospyder_manifest_file)
# Put flextables on the left by default
knitr::opts_chunk$set(ft.align='left')

# now that we have all our params etc loaded, we can set the HTML title of our document 
```
---
title: "`r params$platform` DEG analysis - metadata and statistics"
---


# `r params$project_title` - metadata and statistics {-}

Date report generated: `r format(Sys.time(), '%d %B, %Y')`  

Report prepared by: `r params$project_bioinformatics_name`

Report prepared for: `r params$project_name`


***

This report contains metadata, statistics, and plots related to the experiment.

```{r load_facet_data, include=FALSE}
source(here::here("scripts","load-facet-data.R"), local = TRUE)
```

# Metadata tables {.tabset .tabset-fade}

This table shows the final list of samples that were used in the data analysis (as well as the corresponding sample information, e.g., to which experimental group samples belong).  

This should be reviewed carefully to ensure that samples are correctly identified and included as intended.


**Some samples may have already been removed in study-wide QC.**

## Samples used in this report

```{r metadata-report, collapse=TRUE, eval=TRUE}
# table of sample metadata


set_flextable_defaults(
  font.family = "Arial", font.size = 10, 
  border.color = "gray")

flextable(exp_metadata_subset) %>% 
  add_footer_lines("Samples and corresponding experimental conditions used in this report")
```

## Samples from the experiment excluded in this report

```{r metadata-removed, collapse=TRUE, eval=TRUE}

# table of sample metadata
removed <- exp_metadata %>% dplyr::filter(!original_names %in% exp_metadata_subset$original_names)

flextable(removed) %>%
  add_footer_lines("Samples and removed from this report")
```

## All metadata, as provided

```{r metadata-all, collapse=TRUE, eval=TRUE}

flextable(exp_metadata) %>%
  add_footer_lines("All samples, including removed ones")
```

# Data filtered by the R-ODAF criteria

```{r filtered, eval=TRUE}

# table of R-ODAF filtered genes

if (!is.na(params$display_group_filter)) {
  filtered_table <- filtered_table %>%
    dplyr::filter(facet == params$display_group_filter)
}

flextable(filtered_table) %>%
  add_footer_lines("Number of genes filtered by R-ODAF criteria") %>%
  flextable::autofit()
```

# Contrasts (comparisons)

The experimental comparisons of interest to be tested for this experiment are shown below.

```{r contrasts, eval=TRUE}
# table of contrasts

flextable(contrasts_subset) %>%
  set_header_labels(V1 = "Condition", V2 = "Control") %>%
  add_footer_lines("Contrasts being tested in this analysis") %>%
  flextable::autofit()

# table of #samples/group
nSamples_table <- table(exp_metadata_subset %>% 
                          mutate_all(list(~ str_c(.))) %>%
                          dplyr::pull(!!sym(params$design)))

flextable(as.data.frame(nSamples_table)) %>%
  set_header_labels(Var1 = "Experimental group", Freq = "Number of samples in group") %>%
  add_footer_lines("Number of samples in each experimental group used in this report") %>%
  flextable::autofit()

# table of #DEGs/contrast 
summary_counts <- data.frame()
comparisons <- names(resultsListDEGs)
for(comp in comparisons){ # by comparison
    res <- resultsListDEGs[[comp]]
    if(is.null(res)){
      c <- 0
    } else{
      c <- nrow(res)
    }
    row <- data.frame(comparison=comp, DEG=c)
    summary_counts <- rbind(summary_counts, row)
}

# Does this spit out NA for contrasts where there are no DEGs? It should
flextable(summary_counts) %>%
  set_header_labels(comparison="Contrast", DEG="Number of DEGs") %>%
  add_footer_lines("Number of differentially expressed genes across each contrast")  %>%
  flextable::autofit()
```


# Descriptive plots

## Boxplot of reads per sample group

This section shows the number of reads obtained for sample groups included in the report.  

```{r nReads_boxplot, collapse=TRUE, eval=TRUE}
df<-data.frame(colSums(counts(dds)))
names(df) <- c("nReads")
df$original_names = rownames(df)
df2 <- df %>%
    left_join(exp_metadata) %>%
    dplyr::select(original_names,sym(design_to_use),nReads)

ggplot(df2, aes_string(x=design_to_use, y="nReads") )+
    geom_boxplot() +
    geom_jitter(size = 0.3, alpha = 0.5, color = "blue") +
    coord_flip() +
    expand_limits(y = 0) +
    scale_y_continuous(labels = function(x) format(x, scientific = TRUE)) +
    xlab("Group") +
    ylab("Number of reads") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))

```

## Sample-to-sample distances {.tabset .tabset-fade}

### All genes
```{r sample-to-sample_heatmaps, fig.width = 10, fig.height = 10, eval=TRUE}

plot_sample_heatmap <- function(current_rld){
  ## Obtain the sample euclidean distances
  sampleDists <- stats::dist(t(assay(current_rld)))
  sampleDists <- dendextend::sort_dist_mat(sampleDists)
  sampleDistMatrix <- as.matrix(sampleDists)
  
  distances_df <- as.data.frame(colData(current_rld)[, params$intgroup, drop = F])
  
  ## Define colors to use for the heatmap
  colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
  
  ## Make the heatmap
  return( pheatmap(sampleDistMatrix,
           clustering_distance_rows = sampleDists,
           clustering_distance_cols = sampleDists,
           color = colors,
           annotation_col = distances_df) )
}
# all genes
plot_sample_heatmap(rld)

```

### DEGs only
```{r sample-to-sample_heatmaps_DEGs, fig.width = 10, fig.height = 10, eval=TRUE}
# DEGs only
plot_sample_heatmap(rld_DEGs)
```

### Most `r params$nBest` variable genes only
```{r sample-to-sample_most_variable, fig.width = 10, fig.height = 10, eval=TRUE}
# 100 most variable genes
plot_sample_heatmap(rld_top)
```

## P-value Distributions {.tabset .tabset-fade}

### Distribution of unadjusted p-values
```{r pvalue_distribution, eval=TRUE}
ggplot(data.frame(allResults), aes(x = pvalue)) +
    geom_histogram(alpha = .5, position = 'identity', bins = 50) +
    labs(title = 'Histogram of unadjusted p-values') +
    xlab('Unadjusted p-values') +
    facet_wrap( ~ contrast, ncol = 2) +
    theme_bw()
```


### Distribution of adjusted p-values
```{r adjusted_pvalue_distribution, eval=TRUE}
ggplot(data.frame(allResults), aes(x = padj)) +
    geom_histogram(alpha = .5, position = 'identity', bins = 50) +
    labs(title = 'Histogram of adjusted p-values') +
    xlab('adjusted p-values') +
    facet_wrap( ~ contrast, ncol = 2) +
    theme_bw()
```

### Summary table of (unadjusted) p-value distribution
```{r pvalue_distribution_table, eval=TRUE}
## Split features by different p-value cutoffs
pval_table <- lapply(c(1e-04, 0.001, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5,
    0.6, 0.7, 0.8, 0.9, 1), function(x) {
    data.frame('Cut' = x, 'Count' = sum(allResults$pvalue <= x, na.rm = TRUE))
})
pval_table <- do.call(rbind, pval_table)
#kable(pval_table, format = 'markdown', align = c('c', 'c'))

# knitr::kable(pval_table,
#              row.names =  F,
#              caption = "Summary of the distribution of the p-values") %>%
#   kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive", full_width = F, position = "left")) %>%
#   scroll_box(height = "480px")
flextable(pval_table) %>%
  add_footer_lines("Summary of the distribution of the p-values")
```



## MA plots {.tabset .tabset-fade}

This section contains three groups of MA plots (see [Wikipedia](https://en.wikipedia.org/wiki/MA_plot)) that compare the mean of the normalized counts against the log fold change. Each of the groups has a tab for each contrast. The plots show one point per feature. The points are shown in red if the feature has an adjusted p-value less than the cutoff listed in each section, that is, the statistically significant features are shown in red.

This group of plots shows `alpha` = `r params$alpha`, which is the `alpha` value used to determine which resulting features were significant when running the function `DESeq2::results()`.  

```{r MA_plots, results='asis', eval=TRUE}
foldchangestats <- boxplot(abs(allResults$log2FoldChange), plot = F)
ylims_for_ma_plots <- c(-foldchangestats$stats[4, 1], foldchangestats$stats[4, 1])

for (i in seq_along(resultsListAll)) {
  if(!is.null(resultsListAll[[i]])){
    contrast = gsub(pattern = paste0("log2.*", params$design, "\ "),
                    replacement =  "",
                    x = resultsListAll[[i]]@elementMetadata[[2]][2])
    cat("###", contrast, "  \n\n")
    MA_title <- paste0('MA plot, alpha = ', metadata(resultsListAll[[i]])$alpha,', ', contrast)
    DESeq2::plotMA(resultsListAll[[i]],
                   alpha = metadata(resultsListAll[[i]])$alpha,
                   main = str_wrap(MA_title, width = 40),
                   ylim = ylims_for_ma_plots)
    cat('  \n\n')
  }
}


```

## Standard Deviation vs Mean Counts  {.tabset .tabset-fade}

This section shows how standard deviation changes across the range of counts.

### Regularized counts

```{r SD_plots_log, eval=TRUE}
vsn::meanSdPlot(assay(rld))

```

### log2(n + 1) normalized counts

```{r SD_plots_rld, eval=TRUE}
sd_log <- normTransform(dds)
vsn::meanSdPlot(assay(sd_log))

```


```{r, child=file.path(paths$root,'Rmd','session_info.Rmd')}
```