---
params:
  projectdir: null # Use when loading. If null, here::here() will be used. Can also hard code if for some reason you need that
  project_title: "Test"    # Change depending on your project name
  researcher_name: "Name" # Name of the researcher leading the project
  bioinformatician_name: "Name" # Name of the person running the R-ODAF analyses
  metadata_file: "metadata.txt" # Name of the metadata file. Must be in inputs/metadata/ 
  contrasts_file: "contrasts.txt" # Name of the contrasts file. Must be in inputs/contrasts/ 
  project_description: null # description of project for report, optional
  batch_var: "batch"          # For example, multiple plates
  dose: "dose"               # If there is a dose in the experiment; otherwise use NULL
  solvent_control: "solvent_control"
  platform: "TempO-Seq"      # TempO-Seq Or RNA-Seq
  nmr_threshold: 100000    # 10% of 1M reads for TempOSeq = 100,000; 10% of 10M reads for RNA-Seq = 1,000,000.
  write_additional_output: TRUE # export BMD, biomarker, and biosets output? facet variable should contain chemical and timepoint information
  celltype: "Liver spheroids" # only required for biosets output. Cell type or species used
  units: "uM" # only required for biosets output. Units of dose
  biospyder_dbs: "~/shared/dbs/biospyder/"
  biospyder_manifest_file:  "Human_S1500_Surrogate_2.0_standardized.csv"  
  analysis_name: "default" # short string identifying analysis settings. This will appear in the analysis directory and file names
  species: "human"                 # one of human, mouse, rat, hamster
  design: "Group"                # single experimental group of interest; entries in this column must match the contrast names.
  intgroup: ["Group"]            # experimental group of interest plus covariates; can be more than one
  intgroup_to_plot: ["group"]   # for PCA plots, add tabs for one or more groups to color
  formula_override: null # e.g., "~batch + condition", to enable users to specify the DESeq2 formula to use
  deseq_facet: "Chemical_S9"       # If you have many different experimental groups, you may subset the report by specifying a column in the metadata to filter groups, and then setting the group of interest in deseq_filter
  deseq_filter:  null        # Which group will this report be done on?
  reports_facet: "chemical_timepoint" #  
  reports_filter:  "estradiol 10d" # 
  sortcol: "dose"                # Optionally, a column by which to sort the contrasts
  lenient_contrasts: FALSE         # Use either column (exp, cont) of contrasts file to limit what is included in the report (instead of just exp column)
  strict_contrasts: FALSE         # Use BOTH columns (exp, cont) of contrasts file to limit what is included in the report
  exclude_samples: null      # Optionally, a vector of sample names to exclude from the analysis
  exclude_groups: null       # Optionally, a vector of groups to exclude from the analysis. By default this is assumed to be in the column specified by params$design.
  include_only_column:  null # Restrict analysis to group(s) in the column listed here based on params$include_only_group.
  include_only_group:  null  # Restrict analysis to this/these group(s) within the column listed in params$include_only_column
  cpus: 41                       # Set to a lower number (e.g., 2 to 4) if you aren't working in a server environment
  run_pathway_analysis: TRUE     # Optionally disable pathway analysis if not available for your organism
  wikipathways_directory: "~/shared/dbs/wikipathways"
  linear_fc_filter_DEGs: 1.5 # Default 1.5
  linear_fc_filter_biosets: 1.2 # Default 1.2. Used for biosets output when write_additional_output=True  
  KEGGpathways_filename: "c2.cp.kegg_medicus.v2023.2.Hs.symbols.gmt"
  wikipathways_filename: "wikipathways-20210810-gmt-Homo_sapiens.gmt"
  nBestFeatures: 20              # The number of best features to make plots of their counts
  nBest: 100                     # Number of features to include in table and limiting PCA/clustering analysis
  nHeatmap: 50                   # Number of most variable genes for heatmap
  nHeatmapDEGs: 50               # Number of DEGs for heatmap
  cooks: FALSE                   # the DESeq Cook's distance cutoff, or FALSE to disable it
  filter_gene_counts: FALSE      # Filter genes to those with at least 1 count in 1 sample. Disabled by default for biomarker file output. Can enable to increase speed of analysis
  generate_main_report: TRUE
  generate_stats_report: TRUE
  generate_data_explorer_report: TRUE
  generate_go_pathway_report: TRUE
  generate_tgxddi_report: TRUE
  generate_tgxhdaci_report: TRUE
  generate_summary_report: TRUE
  generate_runningfisher_report: TRUE
  output_digits: 5
  parallel: FALSE
  species_data: null
  count_data_file: null
  sampledata_sep: null
  MinCount: null
  alpha: null
  feature_id: null
  biospyder: null
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    code_folding: hide
    theme: spacelab           # flatly spacelab sandstone cerulean
    code_download: true
bibliography: "`r file.path(params$projectdir, 'references/references.bib')`"
---


```{r load_libraries, message = F, warnings = F, echo = F}
library(here)
library(tidyverse)
library(knitr)
library(kableExtra)
library(ggplot2)
library(formattable)
```

```{r paths}
############################################################################
# File paths in project directory
############################################################################
# paths <- R.ODAF.utils::set_up_filepaths(params)
knitr::opts_knit$set(root.dir = paths[["projectdir"]])
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

```{r setup, include=FALSE}
startTime <- Sys.time()
knitr::opts_chunk$set(warning = FALSE, message = FALSE)

has_description <- !is.na(params$project_description)
description_text <- paste("Purpose of report:", params$project_description)
```


# `r params$project_title` Summary Report {-}

Date report generated: `r format(Sys.time(), '%d %B, %Y')`  

Report prepared by: `r params$bioinformatician_name`

Report prepared for: `r params$researcher_name`

`r if(has_description){description_text}`

***



# Sample Quality Control 

Quality control of each sample is performed according to the criteria laid out in @harrill2021high, plus an additional filter for the distance between samples in hierarchical clustering, suggested by Andrew Williams (Health Canada). The following sections give a high-level overview of the quality control (QC) results for the samples in this study. Much greater detail on filtering criteria for each sample, along with various other QC checks, figures, and tables, can be found in the Studywide Sample QC report.


## Quality control criteria

```{r criteria_table}
# Workaround because "get_params(context = "analysis") leaves out QC parameters
config <- yaml::read_yaml(here::here("inputs", "config", "config.yaml"), eval.expr = TRUE)

qc_pars <- c(config$QC)
qc_pars <- replace_nulls_in_config(qc_pars)

criteria <- data.frame(
  Abbreviation = c("Dendrogram", "Group Dendrogram", "NMR", "FMR", "Ncov5", "Nsig80", "GiC"),
  Description = c("All-samples clustering distance",
                  "Group-specific clustering distance",
                  "Number of mapped reads",
                  "Fraction of uniquely mapped reads", 
                  "Number of probes with at least 5 uniquely mapped reads", 
                  "Number of probes capturing the top 80% of a signal in a sample", 
                  "Gini coefficient"),
  "Default Threshold" = c("< 0.1",
                          "N/A",
                          "> 10% of target read depth",
                          "> 0.5",
                          "within 3*IQR",
                          "within 3*IQR",
                          "RNA-Seq: < 0.99, \nTempO-Seq: < 0.95"),
  "Threshold" = c(
    paste0("< ",qc_pars$studywide_tree_height_cutoff),
    paste0("<", qc_pars$group_tree_height_cutoff),
    paste0("> ",params$nmr_threshold),
    paste0("> ",qc_pars$align_threshold), 
    "within 3*IQR", "within 3*IQR", 
    paste0("< ",qc_pars$gini_cutoff)
  )
)

# Generate and style the table
criteria %>%
  kable("html", caption = "Criteria Table",
        col.names = c("Abbreviation", "Description", "Default Threshold", "Threshold")) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))

```

## Number of samples passing quality control

```{r passing_QC}
metadata <- R.ODAF.utils::get_metadata(file.path(paths$metadata, params$metadata_file), paths)

tech_ctrl_names <- metadata %>%
  dplyr::filter(.data[[qc_pars$technical_control]] == "T" | .data[[qc_pars$technical_control]] == "TRUE") %>%
  dplyr::pull(original_names)

sample_names <- metadata %>%
  dplyr::filter(.data[[qc_pars$technical_control]] == "F" | .data[[qc_pars$technical_control]] == "FALSE") %>%
  dplyr::filter(.data[[params$solvent_control]] == "F" | .data[[params$solvent_control]] == "FALSE") %>%
  dplyr::pull(original_names)

solvent_names <- metadata %>%
  dplyr::filter(.data[[qc_pars$technical_control]] == "F" | .data[[qc_pars$technical_control]] == "FALSE") %>%
  dplyr::filter(.data[[params$solvent_control]] == "T" | .data[[params$solvent_control]] == "TRUE") %>%
  dplyr::pull(original_names)

ref_samples <- metadata %>%
  dplyr::filter(.data[[qc_pars$reference_rna]] == "T" | .data[[qc_pars$reference_rna]] == "TRUE") %>%
  dplyr::pull(original_names)

qc_metadata <- R.ODAF.utils::get_metadata(file.path(paths$qc, "metadata.QC_applied.txt"), paths)

qc_info <- file.path(paths$qc, "details", "QC_per_sample.txt")
qc_info <- read.delim(qc_info, sep = "\t", stringsAsFactors = FALSE)

passed <- qc_info %>%
  dplyr::filter(Any == "PASS") %>%
  dplyr::pull(sample_ID)

num_passed_refs <- passed %in% ref_samples %>% sum()
num_passed_samples <- passed %in% sample_names %>% sum()
num_passed_solvent <- passed %in% solvent_names %>% sum()
num_passed_total <- num_passed_samples + num_passed_solvent

pass_table <- data.frame(
  "Sample Type" = c("Experimental Samples", "Solvent controls","Reference RNA", "Technical Controls"),
  "Total Number" = c(length(sample_names), length(solvent_names), length(ref_samples), length(tech_ctrl_names)),
  "Number Passing QC Filters" = c(num_passed_samples, num_passed_solvent, "N/A*", "N/A*")
)

pass_table %>%
  kable("html", caption = "Passing QC Table",
        col.names = c("Sample Type", "Total number", "Number passing QC Filters")) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```
*Note: Technical controls and reference RNA are used to check the quality of the sequencing run and to compare between batches. They are not subject to QC filters.

```{r QC_fails}
fail_dendro <- qc_info %>%
  dplyr::filter(dendrogram == "FAIL") %>%
  dplyr::select(sample_ID, params$dendro_color_by)

fail_groupdendro <- qc_info %>%
  dplyr::filter(group_dendrogram == "FAIL") %>%
  dplyr::select(sample_ID, params$dendro_color_by)

fail_nmr <- qc_info %>% 
  dplyr::filter(NMR == "FAIL") %>%
  dplyr::select(sample_ID, params$dendro_color_by)

fail_fmr <- qc_info %>% 
  dplyr::filter(FMR == "FAIL") %>%
  dplyr::select(sample_ID, params$dendro_color_by)

fail_ncov5 <- qc_info %>% 
  dplyr::filter(Ncov5 == "FAIL") %>%
  dplyr::select(sample_ID, params$dendro_color_by)

fail_nsig80 <- qc_info %>% 
  dplyr::filter(Nsig80 == "FAIL") %>%
  dplyr::select(sample_ID, params$dendro_color_by)

fail_gic <- qc_info %>% 
  dplyr::filter(Gini == "FAIL") %>%
  dplyr::select(sample_ID, params$dendro_color_by)
```

## Samples Failing Quality Control {.tabset .tabset-pills}


### Clustering Distance (studywide)
```{r dendro_fail}
fail_dendro %>%
  kable("html", caption = "Samples Failing Clustering Distance Filter") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```

<br>
<br>

### Clustering Distance (by group)
```{r group_dendro_fail}
fail_groupdendro %>%
  kable("html", caption = "Samples Failing By-Group Clustering Distance Filter") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```

<br>
<br>

### Number of Mapped Reads
```{r nmr_fail}
fail_nmr %>%
  kable("html", caption = "Samples Failing Number of Mapped Reads Filter") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```

<br>
<br>

### Fraction of Uniquely Mapped Reads
```{r fmr_fail}
fail_fmr %>%
  kable("html", caption = "Samples Failing Fraction of Uniquely Mapped Reads Filter") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))

```

<br>
<br>

### Number of Probes with at Least 5 Uniquely Mapped Reads
```{r ncov5_fail}
fail_ncov5 %>%
  kable("html", caption = "Samples Failing Number of Probes with at Least 5 Uniquely Mapped Reads Filter") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```

<br>
<br>

### Number of Probes Capturing the Top 80% of a Signal in a Sample
```{r nsig80_fail}
fail_nsig80 %>%
  kable("html", caption = "Samples Failing Number of Probes Capturing the Top 80% of a Signal in a Sample Filter") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```

<br>
<br>

### Gini Coefficient
```{r gic_fail}
fail_gic %>%
  kable("html", caption = "Samples Failing Gini Coefficient Filter") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```


<br>
<br>
   
  
  

# Differential Expression

```{r }
read_type <- if (params$platform == "TempO-Seq") {
  "probes"
} else {
  "genes"
}
```

A count matrix containing the $`r num_passed_total`$ samples that passed QC was imported into R for statistical analysis [@rcitation]. Following the recommendations set out by the Omics Data Analysis Frameworks for Regulatory application (R-ODAF) guidelines [@VERHEIJEN2022105143], `r read_type` were filtered to include only those where 75% of at least one experimental group were above `r params$MinCount` CPM, and spurious spikes were removed in which (max - median) of counts were less than (sum of counts)/(number of replicates + 1). We used DESeq2 `r packageVersion("DESeq2")` [@love2014deseq2] to test for differentially abundant genes within the `r params$platform` data. The log2FoldChange shrinkage procedure used was ashr [@stephens2017false]. An alpha of `r params$alpha` was used to extract raw results, which are reported as the `r gsub(mcols(x=overallResListAll[[1]][[1]])$description[[4]],pattern=":.*",replacement="")`. To account for multiple testing, `r mcols(overallResListAll[[1]][[1]])$description[5]` are reported. Cook's cutoff was set to `r params$cooks` in this analysis (a value of FALSE indicates that Cook's distance filtering was disabled). Differentially expressed genes (DEGs) were filtered using a linear fold change cutoff of `r params$linear_fc_filter_DEGs` and adjusted p-value of `r params$alpha`.



```{r DEGs}
# read in contrasts
exp_contrasts <- R.ODAF.utils::get_contrasts(file.path(paths$contrasts, params$contrasts_file), paths)

mixedrank <- function(x) order(gtools::mixedorder(x))
deg_counts <- read.delim(file.path(paths$summary, "DEG_counts.txt"), 
                         sep = "\t", stringsAsFactors = FALSE, header = TRUE) %>%
  arrange(mixedrank(comparison))

if (!is.na(params$reports_facet)) {
  colnames(deg_counts) <- c(params$reports_facet, "Contrast", "DEGs", "Downregulated", "Upregulated")
  deg_table_alignment <- c("l", "l", "l", "r", "l")
} else {
  colnames(deg_counts) <- c("Contrast", "DEGs", "Downregulated", "Upregulated")
  deg_table_alignment <- c("l", "l", "r", "l")
}
```

There were **`r nrow(exp_contrasts)` contrasts** analyzed in this project. 

Of these, **`r nrow(deg_counts)` contrasts** had differentially expressed genes (DEGs) that passed all R-ODAF filters.

The number of significant DEGs (i.e., those passing all filters) in each contrast is shown in the table and figure below.

## Number of DEGs per Contrast

```{r DEG_table, }

# From https://github.com/renkun-ken/formattable/issues/106#issuecomment-792383056
bg = function(start, end, color, ...) {
  paste("linear-gradient(90deg,transparent ",percent(start),",",
        color, percent(start), ",", color, percent(end),
        ", transparent", percent(end),")")
}

# Colour bar with right-aligned text and bar
color_bar2 =  function (color = "lightgray", fun = "proportion", ...) 
{
    fun <- match.fun(fun)
    formatter("span", style = function(x) style(display = "inline-block",
                `unicode-bidi` = "plaintext", 
                "background" = bg(1-fun(as.numeric(x), ...), 1, color), "width"="100%" ))
}

# Colour bar with left-aligned text and bar
color_bar3 = function(color = "lightgray", fun = "proportion", ...) {
  fun <- match.fun(fun)
  formatter("span", style = function(x) {
    value <- fun(as.numeric(x), ...)
    style(display = "inline-block",
          `unicode-bidi` = "plaintext",
          "background" = bg(0, value, color),
          "width" = "100%",
          "text-align" = "left")
  })
}

deg_counts %>%
  formattable(
              align = deg_table_alignment,
              list(DEGs = color_tile("white", "lightgreen"),
                   Upregulated = color_bar3("pink"),
                   Downregulated = color_bar2("lightblue"))) %>%
  formattable::as.htmlwidget(width = 500)
```

<br>
<br>

```{r DEG_figures}
# facets <- R.ODAF.utils::get_facets(qc_metadata, params)

if (!is.na(params$reports_facet)) {
  show_deg_plots <- TRUE
  
  plot_list <- summarize_across_facets(overallResListAll, overallResListDEGs, filtered_table, facets, params)

  p1 <- plot_list$p1
  p2 <- plot_list$p2

} else {
  show_deg_plots <- FALSE
}

if (show_deg_plots == TRUE) {
  p1
}

# summarize_across_facets errors when reports_facet = NA

```

## DEGs filtered by R-ODAF criteria
```{r p2, fig.height=12}
if (show_deg_plots == TRUE) {
  p2
}
```




# Biomarkers
Biomarker analysis is **only available on human datasets** and the biomarkers available in this workflow have only been tested on specific cell lines. 
The results of this analysis may not be applicable to other cell lines or tissues.

The TGx-DDI biomarker was developed using the **TK6** cell line [@li2015development] and has been tested on **HepaRG** [@buick2020tgxddi_hepaRG].

The TGx-HDACi biomarker was developed using the **TK6** cell line [@cho_hdaci_2019] and has not been tested on other cells.

<span style="color:red;">Cell line used in this study: `r params$celltype` </span>

```{r check_biomarkers}
if (params$generate_tgxddi_report && file.exists(file.path(paths$summary, "tgx-ddi_results_summary.csv"))) {
  ddi_run <- TRUE

  ddi_results <- read.csv(file.path(paths$summary, "tgx-ddi_results_summary.csv"), stringsAsFactors = FALSE)

  ddi <- ddi_results %>%
    dplyr::select(ID, Overall) %>%
    rename(DDI_result = Overall)
} else {
  ddi_run <- FALSE
}

if (params$generate_tgxhdaci_report && file.exists(file.path(paths$summary, "tgx-hdaci_results_summary.csv"))) {
  hdaci_run <- TRUE

  hdaci_results <- read.csv(file.path(paths$summary, "tgx-hdaci_results_summary.csv"), stringsAsFactors = FALSE)

  hdaci <- hdaci_results %>%
    dplyr::select(ID, Overall) %>%
    rename(HDACi_result = Overall)
} else {
  hdaci_run <- FALSE
}

# If ddi_run is TRUE and hdaci_run is TRUE, merge tables on ID
if (ddi_run && hdaci_run) {
  biomarker_table <- merge(ddi, hdaci, by = "ID", all = TRUE)
} else if (ddi_run && !hdaci_run) {
  biomarker_table <- ddi
} else if (hdaci_run && !ddi_run) {
  biomarker_table <- hdaci
} else if (!ddi_run && !hdaci_run) {
  biomarker_table <- NULL
}

ddi_formatter <- formatter("span",
                           style = x ~ ifelse(x == "DDI", 
                                              style("background-color" = "pink"), 
                                              style("background-color" = "lightblue")))

hdaci_formatter <- formatter("span",
                             style = x ~ ifelse(x == "HDACi", 
                                                style("background-color" = "pink"), 
                                                style("background-color" = "lightblue")))

# If biomarker_table is not NULL, print the table
if (!is.null(biomarker_table)) {
  biomarker_table %>%
    rename(Group = ID) %>%
    arrange(mixedrank(Group)) %>%
    formattable(list(DDI_result = ddi_formatter, HDACi_result = hdaci_formatter)) %>%
    formattable::as.htmlwidget()
} else {
  print("No biomarker analyses were run.")
}
```








## References {-}

If these methods were used in your study, please cite the following papers as appropriate:  



<div id="refs"></div>

# Session Info
```{r, child=file.path(paths$projectdir,'Rmd','session_info.Rmd')}
```